Question 3

Use dataset available on http://users.stat.ufl.edu/~winner/data/nfl2008_fga.csv

* Use LDA to classify the dataset into few classes so that at least 90% of information of dataset is explained through new classification. (Hint: model the variable “qtr” to variables “togo”, “kicker”, and “ydline”). How many LDs do you choose? Explain the reason.
* Apply PCA, and identify the important principle components involving at least 90% of dataset variation. Explain your decision strategy?  Plot principle components versus their variance.

In [111]:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [112]:
import pandas as pd
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB
from sklearn.decomposition import PCA
import numpy as np
import time

In [113]:
nfl_df = pd.read_csv("http://users.stat.ufl.edu/~winner/data/nfl2008_fga.csv")
nfl_df.head()

GameDate AwayTeam HomeTeam  qtr  min  sec kickteam  def  down  togo  ...  \
0  20081130      IND      CLE    1   47    2      IND  CLE   4.0  11.0  ...   
1  20081005      IND      HOU    1   54   47      IND  HOU   4.0   3.0  ...   
2  20081228      TEN      IND    1   45   20      IND  TEN   4.0   3.0  ...   
3  20081012      BAL      IND    1   45   42      IND  BAL   4.0   1.0  ...   
4  20080907      CHI      IND    1   50   56      IND  CHI   4.0  21.0  ...   

   distance  homekick kickdiff  timerem  offscore  defscore  season  GOOD  \
0        30         0       -3     2822         0         3    2008     1   
1        46         0        0     3287         0         0    2008     1   
2        28         1        7     2720         7         0    2008     1   
3        37         1       14     2742        14         0    2008     1   
4        39         1        0     3056         0         0    2008     1   

   Missed  Blocked  
0       0        0  
1       0        0  
2       0        0  
3       0        0  
4       0        0  

[5 rows x 23 columns]

In [114]:
nfl_df["qtr"].unique()

array([1, 2, 3, 4, 5], dtype=int64)

In [115]:
nfl_df.isna().sum()

GameDate    0
AwayTeam    0
HomeTeam    0
qtr         0
min         0
sec         0
kickteam    0
def         0
down        2
togo        2
kicker      0
ydline      0
name        0
distance    0
homekick    0
kickdiff    0
timerem     0
offscore    0
defscore    0
season      0
GOOD        0
Missed      0
Blocked     0
dtype: int64

In [116]:
nfl_df = nfl_df.fillna(0)
nfl_df.isna().sum()

GameDate    0
AwayTeam    0
HomeTeam    0
qtr         0
min         0
sec         0
kickteam    0
def         0
down        0
togo        0
kicker      0
ydline      0
name        0
distance    0
homekick    0
kickdiff    0
timerem     0
offscore    0
defscore    0
season      0
GOOD        0
Missed      0
Blocked     0
dtype: int64

In [117]:
nfl_df = nfl_df.drop(["GameDate","AwayTeam","HomeTeam","kickteam","def", "name"], axis=1)

In [118]:
nfl_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1039 entries, 0 to 1038
Data columns (total 17 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   qtr       1039 non-null   int64  
 1   min       1039 non-null   int64  
 2   sec       1039 non-null   int64  
 3   down      1039 non-null   float64
 4   togo      1039 non-null   float64
 5   kicker    1039 non-null   int64  
 6   ydline    1039 non-null   int64  
 7   distance  1039 non-null   int64  
 8   homekick  1039 non-null   int64  
 9   kickdiff  1039 non-null   int64  
 10  timerem   1039 non-null   int64  
 11  offscore  1039 non-null   int64  
 12  defscore  1039 non-null   int64  
 13  season    1039 non-null   int64  
 14  GOOD      1039 non-null   int64  
 15  Missed    1039 non-null   int64  
 16  Blocked   1039 non-null   int64  
dtypes: float64(2), int64(15)
memory usage: 138.1 KB


In [119]:
X = nfl_df.drop("qtr", axis=1)
y = nfl_df["qtr"]

In [120]:
clfs = {
    "DT" : DecisionTreeClassifier(),
    "LR" : LogisticRegression(),
    "GB" : GaussianNB()
}

In [121]:
accuracy_normal = {}
y_preds = {}
for key in clfs:
    for i in range(1000):
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state= 42)
        clf = clfs[key]
        t_clf = clf.fit(X_train, y_train)
        y_pred = t_clf.predict(X_test)
        a = accuracy_score(y_test, y_pred)
        accuracy_normal[key]= a
        y_preds[key]= y_preds
pd.DataFrame([accuracy_normal])

DT        LR        GB
0  1.0  0.996795  0.945513

#### Using LDA

LDA Creates a linier combinations of the original features

`n_componets = 0.5 * x1 + 0.2 * x2 + 0.3 *x3`

In [122]:
lda_based_model = LinearDiscriminantAnalysis(n_components=1)
lda_X = lda_based_model.fit_transform(X, y)

In [123]:
var_ratio_lda = lda_based_model.explained_variance_ratio_
var_ratio_lda

array([0.99244456])

#### Creating a function to select n_components

In [124]:
def n_components(var_ratio_lda, des_var) -> int:
    variance = 0.0
    n_comp = 0
    for var_ratio in var_ratio_lda:
        variance = variance + var_ratio
        n_comp += 1
        if variance >= des_var:
            break
    return n_comp

In [125]:
comp = n_components(var_ratio_lda, 0.90)
comp

1

In [126]:
lda_df = pd.DataFrame(lda_X)
lda_df.columns = ['ldaOne']

In [127]:
accuracy_lda = {}
y_preds = {}
time_taken = {}
for key in clfs:
    for i in range(1000):
        start_time = time.time()
        X_train, X_test, y_train, y_test = train_test_split(lda_X, y, test_size = 0.3, random_state= 42)
        clf = clfs[key]
        t_clf = clf.fit(X_train, y_train)
        y_pred = t_clf.predict(X_test)
        end_time = time.time()
        a = accuracy_score(y_test, y_pred)
        accuracy_lda[key]= a
        y_preds[key]= y_preds
        time_taken[key] = np.subtract(end_time, start_time)
warnings.filterwarnings("ignore")
pd.DataFrame([accuracy_lda])

DT        LR        GB
0  0.980769  0.971154  0.967949

As we can seeLiner Regression takes More time than other models and as we have seen above it also has lesser accuarcy than decision tree, hence we will go with DecisionTreeClassifier

In [128]:
pd.DataFrame([time_taken])

DT        LR     GB
0  0.000998  0.022999  0.001

PCA

In [129]:
pca_ins = PCA(n_components=1)
X_pca = pca_ins.fit_transform(X)
PCA_L = pca_ins.components_
PCA_L

array([[-1.66483606e-02, -9.23988619e-04, -6.95769246e-05,
         1.01183727e-04,  3.71613583e-04,  4.69406766e-04,
         4.75019033e-04,  1.03303163e-06,  9.21338193e-04,
        -9.99825622e-01,  6.32715643e-03,  5.40581823e-03,
         0.00000000e+00, -3.14239886e-06,  3.14239886e-06,
         2.32335195e-06]])

In [130]:
pca_df = pd.DataFrame(PCA_L)
pca_df_T =  pca_df.T
pca_df_T.columns = ["pca"]
pca_df_T.head()

pca
0 -0.016648
1 -0.000924
2 -0.000070
3  0.000101
4  0.000372

In [131]:
pca_ins.explained_variance_, pca_ins.n_components

(array([1007908.25197757]), 1)

In [132]:
X.shape

(1039, 16)

In [133]:
df_final = np.dot(X, pca_df_T)

In [134]:
accuracy_pca = {}
y_preds = {}
time_taken = {}
for key in clfs:
    for i in range(1000):
        start_time = time.time()
        X_train, X_test, y_train, y_test = train_test_split(df_final, y, test_size = 0.3, random_state= 42)
        clf = clfs[key]
        t_clf = clf.fit(X_train, y_train)
        y_pred = t_clf.predict(X_test)
        end_time = time.time()
        a = accuracy_score(y_test, y_pred)
        accuracy_pca[key]= a
        y_preds[key]= y_preds
        time_taken[key] = np.subtract(end_time, start_time)
warnings.filterwarnings("ignore")
pd.DataFrame([accuracy_pca])

DT        LR        GB
0  1.0  0.996795  0.983974

#### Conclusion

#### Accuracy Normal 1000 MC runs

In [135]:
pd.DataFrame([accuracy_normal])

DT        LR        GB
0  1.0  0.996795  0.945513

#### Accuracy LDA 1000 MC runs

In [136]:
pd.DataFrame([accuracy_lda])

DT        LR        GB
0  0.980769  0.971154  0.967949

#### Accuracy PCA 1000 MC runs

In [137]:
pd.DataFrame([accuracy_pca])

DT        LR        GB
0  1.0  0.996795  0.983974

The summray of the performance is as shown below

|     Methods  | Decision Tree | Logistic Regression | Naive Bayes |
|--------------|---------------|---------------------|-------------|
|      Normal  | 1.0           | 0.99                | 0.94        |
|      LDA     | 0.980         | 0.97                | 0.96        |
|      PCA     | 0.999         | 0.99                | 0.98        |

n_components for LDA and PCA is 1 for variance of more than 0.96. Increase in value of n_components will train the model faster but will reduce the accuracy